src/framework/components/esmscript/attribute-utils.js

```diff
@@ -0,0 +1,359 @@
+import { GraphNode } from "../../../scene/graph-node.js";
+import { Debug } from "../../../core/debug.js";
+
+/**
+ * @callback UpdateFunction
+ * @param {number} dt - The time since the last update.
+ * @ignore
+ */
+
+/**
+ * @callback SwapFunction
+ * @param {Object} newState - The new state to swap to.
+ * @ignore
+ */
+
+/**
+ * @typedef {Object} ModuleInstance
+ * @property {Function} [active] - A function called when the module becomes active
+ * @property {Function} [inactive] - A function called when the module becomes inactive
+ * @property {UpdateFunction} [update] - A function called on game tick if the module is enabled
+ * @property {Function} [destroy] - A function called when the module should be destroyed
+ */
+
+/**
+ * This type represents a generic class constructor.
+ * @typedef {Function} ModuleClass
+ */
+
+/**
+ * @typedef {Object|Map} AttributeDefinition
+ * @property {'asset'|'boolean'|'curve'|'entity'|'json'|'number'|'rgb'|'rgba'|'string'|'vec2'|'vec3'|'vec4'} type - The attribute type
+ */
+
+/**
+ * The expected output of an ESM Script file. It contains the class definition and the attributes it requires.
+ * @typedef {Object} ModuleExport
+ * @property {ModuleClass} default - The default export of a esm script that defines a class
+ * @property {Object.<string, AttributeDefinition>} attributes - An object containing the names of attributes and their definitions;
+ */
+
+/**
+ * A list of valid attribute types
+ * @ignore
+ */
+export const VALID_ATTR_TYPES = new Set([
+    "asset",
+    "boolean",
+    "curve",
+    "entity",
+    "json",
+    "number",
+    "rgb",
+    "rgba",
+    "string",
+    "vec2",
+    "vec3",
+    "vec4"
+]);
+
+/**
+ * For any given attribute definition returns whether it conforms to the required
+ * shape of an attribute definition.
+ *
+ * @param {AttributeDefinition|object} attributeDefinition - The attribute to check
+ * @returns {boolean} True if the object can be treated as a attribute definition
+ * @example
+ * isValidAttributeDefinition({ type: 'entity' }); // true
+ * isValidAttributeDefinition({ type: 'invalidType' }); // false
+ * isValidAttributeDefinition({ x: 'y' }); // false
+ */
+export const isValidAttributeDefinition = (attributeDefinition) => {
+    return attributeDefinition && attributeDefinition.hasOwnProperty('type');
+};
+
+/**
+ * This function returns the value at a path from an object.
+ * @param {Object} object - The object to get the value from
+ * @param {Array.<string>} path - The path to the value as an array of strings
+ * @returns {*} The value at the path or undefined if it doesn't exist
+ */
+export const getValueAtPath = (object, path) => {
+    return path.reduce((prev, curr) => prev?.[curr], object);
+};
+
+/**
+ * This function sets the value at a path on an object.
+ * @param {Object} object - The object to set the value on
+ * @param {Array.<string>} path - The path to the value as an array of strings
```

Comment:
Convention in repo is `string[]`

Reply:
Thanks @kungfooman, I've changed this, but also noticed we do have some `Array<*>` in other places.